src/dolphin/interpolation.py

```diff
@@ -0,0 +1,149 @@
+        wrapped phase at non-ps pixels.
+
+    References
+    ----------
+    "A persistent scatterer interpolation for retrieving accurate ground
+    deformation over InSAR\-decorrelated agricultural fields"
+    Chen et al., 2015, https://doi.org/10.1002/2015GL065031
+
+    """
+    nrow, ncol = weights.shape
+
+    if np.all(
+        np.logical_or(
+            np.logical_or(weights == 0, weights == 1),
+            np.logical_or(weights is True, weights is False),
+        )
+    ):
+        logger.info("Binary weights, using PS-like interpolation.")
+    else:
+        logger.info(
+            f"Range of values as weights, using weight_cutoff = {weight_cutoff}"
+        )
+
+    # Ensure weights are between 0 and 1
+    if np.any(weights.astype(np.float32) > 1):
+        logger.warning("weights array has values greater than 1. Clipping to 1.")
+    if np.any(weights.astype(np.float32) < 0):
+        logger.warning("weights array has negative values. Clipping to 0.")
+    # Make shared versions of the input arrays to avoid copying in each thread
+    weights_float = np.clip(weights.astype(np.float32), 0, 1)
+
+    interpolated_ifg = np.zeros((nrow, ncol), dtype=np.complex64)
+
+    indices = np.array(get_circle_idxs(max_radius, min_radius=min_radius))
+
+    _interp_loop(
+        ifg,
+        weights_float,
+        weight_cutoff,
+        num_neighbors,
+        alpha,
+        indices,
+        interpolated_ifg,
+    )
+    return interpolated_ifg
+
+
+@numba.njit(parallel=True)
+def _interp_loop(
+    ifg, weights, weight_cutoff, num_neighbors, alpha, indices, interpolated_ifg
+):
+    nrow, ncol = weights.shape
+    nindices = len(indices)
+    for r0 in numba.prange(nrow):
+        # convert linear idx to row, col
+        for c0 in range(ncol):
+            if weights[r0, c0] >= weight_cutoff:
+                interpolated_ifg[r0, c0] = ifg[r0, c0]
+                continue
+
+            csum = 0.0 + 0j
+            counter = 0
+            r2 = np.zeros(num_neighbors, dtype=np.float64)
+            cphase = np.zeros(num_neighbors, dtype=np.complex128)
+
+            for i in range(nindices):
+                idx = indices[i]
+                r = r0 + idx[0]
+                c = c0 + idx[1]
+
+                if (
+                    (r >= 0)
+                    and (r < nrow)
+                    and (c >= 0)
+                    and (c < ncol)
+                    and weights[r, c] >= weight_cutoff
+                ):
+                    # calculate the square distance to the center pixel
+                    r2[counter] = idx[0] ** 2 + idx[1] ** 2
+
+                    cphase[counter] = np.exp(1j * np.angle(ifg[r, c]))
+                    counter += 1
+                    if counter >= num_neighbors:
+                        break
+
+            # TODO : why use the "counter - 1" here to normalize?
+            r2_norm = (r2[counter - 1] ** alpha) / 2
```

Comment:
It looks like the intrinsic assumption here is that the last element of r2 is the largest element. Is that correct assumption? if the largest element is needed why not just get the max of r2? 
I think I do not understans the arbitrary division by 2 here either. I don't know if you or @scottstanie or @taliboliver can help me understand what is going on here.

Reply:
Hi @hfattahi yes I think using max of r^2 should be equivalent and about the division by two my understanding was that is part of the scaling factor from Ke Wang implementation (https://github.com/UT-Radar-Interferometry-Group/psinterp/blob/91b5e797f7768766dfb9ab941ba47cdba479472c/src/cpp/psinterp.cpp#L117) I will take a deeper look into it and add more info on this.

Reply:
![image](https://github.com/isce-framework/dolphin/assets/8291800/382a2dcb-2a93-491c-a8e1-ef83bc0eada6)
1. We could add a reference like `See Eq. (2), Chen 2015`


2. Thanks for asking about the r2, which I tried to explain more [here in a comment](https://github.com/mirzaees/dolphin/pull/1/files#diff-4802b88e403e32ee9c9c7dc279fd6dea658ecbb460a57149addc299b9826a0d3R129-R130) 
But your question made me realize that I did make a change from Ke's code in generating the circle indexes:
https://github.com/mirzaees/dolphin/blob/7115a57e8861766b04f949eb0bd8fa1ceaa02267/src/dolphin/similarity.py#L241-L243

If we know we are going to access all the pixels in the circle
![image](https://github.com/isce-framework/dolphin/assets/8291800/bfcf947e-3c27-4aec-96f7-6d234c47ab74)
then sorting them leads to faster runtime. 

BUT that breaks this assumption here, as his code produced the indexes in a spiral outward so the last one was always at the max radius $R$. So we should probably change the code to `get_circle_idxs(... sort_output=False)` for this function

Reply:
Thanks for the clarification. Let me know if the "get_circle_idxs" is already fixed so that we resolve this comment.

Reply:
Hi @hfattahi, yes it is fixed now. We applied all you suggestions, please take another look so we can merge